presto-main/src/main/java/com/facebook/presto/sql/planner/PlanNodeCanonicalInfo.java

```diff
@@ -26,19 +29,25 @@
 import java.util.Optional;
 
 import static com.facebook.presto.cost.HistoryBasedPlanStatisticsManager.historyBasedPlanCanonicalizationStrategyList;
+import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.graph.Traverser.forTree;
 import static java.util.Objects.requireNonNull;
 
 public class PlanNodeCanonicalInfo
 {
     private final String hash;
-    private final List<PlanStatistics> inputTableStatistics;
+    private final Optional<List<PlanStatistics>> inputTableStatistics;
+    private final Optional<List<InputTableCacheKey>> inputTableCacheKeys;
```

Comment:
We need to cache the InputTableCacheKey so that at the end of query completion we can fetch the corresponding input statistics.